### PR TITLE
[Bindgen]: Add rust wallet core rust library + Aptos blind signing

### PR DESCRIPTION
**Pull request by Milerius**
## Description

Proof of concept for future rust bindgen

## Tasks

- [x] rust-bindgen script
- [x] update CI
- [x] C++ unit tests
- [x] C unit tests
- [x] Swift unit tests
- [x] Kotlin unit tests
- [x] Wasm unit tests
- [x] Link the correct library simulator/universal in swift project
- [x] Verify binary build in the apps

## How to test

- Run C++ unit tests
- Run Swift unit tests
- Run Kotlin unit tests

### CODE REVIEW

File: swift/project.yml [hunk 1]
```diff
@@ -6,6 +6,7 @@ settings:
   base:
     HEADER_SEARCH_PATHS: $(SRCROOT)/wallet-core ${SRCROOT}/trezor-crypto/crypto
     SYSTEM_HEADER_SEARCH_PATHS: ${SRCROOT}/include ${SRCROOT}/../build/local/include ${SRCROOT}/trezor-crypto/include $(SRCROOT)/protobuf /usr/local/include /opt/homebrew/include
+    LIBRARY_SEARCH_PATHS: "$(inherited) ${SRCROOT}/../build/local/aarch64-apple-ios-sim/release"
```
Comment (Milerius) [inline swift/project.yml:9]: Need to find a conditional way to gen for simulator and for ios: https://gist.github.com/surpher/bbf88e191e9d1f01ab2e2bbb85f9b528#cross-compiling

File: bindgen/rust/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+[package]
+name = "walletcore-crypto-bindgen"
```
Comment (hewigovens) [inline bindgen/rust/Cargo.toml:2]:
```suggestion
name = "wallet-core-rs"
```

File: bindgen/rust/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,4 @@
+#[no_mangle]
+pub extern fn rust_function() -> i32 {
```
Comment (hewigovens) [inline bindgen/rust/src/lib.rs:2]: Should we move source code to `src` folder?

File: tools/rust-bindgen [hunk 4]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+
+rustup target add aarch64-linux-android i686-linux-android x86_64-apple-ios aarch64-apple-ios-sim aarch64-apple-ios wasm32-unknown-emscripten
+cargo install cargo-lipo
```
Comment (hewigovens) [inline tools/rust-bindgen:4]: Move to `tools/install-dependencies`

**Review by hewigovens — COMMENTED** [review]
🚀, we may also need to compare final binary size later

File: .github/workflows/linux-sampleapp-ci.yml [hunk 5]
```diff
@@ -42,6 +42,8 @@ jobs:
     - name: Build
       run: |
         make -Cbuild -j12 TrustWalletCore
+        echo "nm"
+        nm build/libTrustWalletCore.a | grep "rust_function"
```
Comment (hewigovens) [inline .github/workflows/linux-sampleapp-ci.yml:46]: can be removed now
```suggestion
```

**Review by hewigovens — COMMENTED** [review]
now we can add non-trivial rust methods

**Comment by miloserdow** [comment]
We should use git submodules when adding Rust/Go libraries from somewhere

**Comment by Milerius** [comment]
> We should use git submodules when adding Rust/Go libraries from somewhere

I do not think that's a good idea for a few reasons:

1. We are already a mono-repo which tends not to use git submodules
2. Git submodules is a hassle to keep updated
3. Rust crate does not use git submodule, considered as counter-intuitive - instead, you can specify a revision of a specific commit
4. Same answer for golang

File: bindgen/rust/src/move_parser/mod.rs [hunk 6]
```diff
@@ -0,0 +1,84 @@
+use std::{
```
Comment (Milerius) [inline bindgen/rust/src/move_parser/mod.rs:1]: Need to add a free function to avoid memory leak

File: bindgen/rust/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+[package]
```
Comment (hewigovens) [inline bindgen/rust/Cargo.toml:1]: Can you commit `Cargo.lock`?

File: include/TrustWalletCore/TWCryptoBindgen.h [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+#include "TWBase.h"
+
+TW_EXTERN_C_BEGIN
+
+TW_EXPORT_STRUCT
+struct TWCryptoBindgen;
+
+TW_EXPORT_STATIC_METHOD
+int32_t TWCryptoBindgenRustFunction() TW_VISIBILITY_DEFAULT;
```
Comment (hewigovens) [inline include/TrustWalletCore/TWCryptoBindgen.h:17]: Let's change this file to `TWHexCoding`

File: rust/src/move_parser/mod.rs [hunk 9]
```diff
@@ -0,0 +1,91 @@
+use std::{
```
Comment (hewigovens) [inline rust/src/move_parser/mod.rs:1]: can you add license header?

**Review by catenocrypt — APPROVED** [review]

**Review by miloserdow — APPROVED** [review]

**Review by miloserdow — DISMISSED** [review]

**Review by miloserdow — APPROVED** [review]

**Review by MaximPestryakov — APPROVED** [review]

**Review by ArtemGoryunov — APPROVED** [review]

**Pull request merged**